File: TaskDistributionTest/TDT.Client/WorkerPool.cs
using Grpc.Core;
using MathCalc.OneArg;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using TIPC.Core.Tools.Threading;

namespace TDT.Client
{
    public class WorkerPool : IDisposable
    {
        Process[] _processCollection { get; }
        Channel[] _channelCollection { get; }
        MathCalc.OneArg.MathCalc.MathCalcClient[] _mathClients { get; }
        int _maxBatchSize { get; }

        public WorkerPool(FileInfo ServerPath, int StartPort, int WorkerCount, int MaxBatchSize)
        {
            var currentPort = StartPort;
            _maxBatchSize = MaxBatchSize;
            _channelCollection = new Channel[WorkerCount];
            _processCollection = new Process[WorkerCount];
            _mathClients = new MathCalc.OneArg.MathCalc.MathCalcClient[WorkerCount];
            for (int i = 0; i < WorkerCount; ++i)
            {
                var port = currentPort++;
                var args = $"-p {port}";
                var proc = Process.Start(ServerPath.FullName, args);
                _processCollection[i] = proc;
                _channelCollection[i] = new Channel($"127.0.0.1:{port}", ChannelCredentials.Insecure);
                _mathClients[i] = new MathCalc.OneArg.MathCalc.MathCalcClient(_channelCollection[i]);
            }
        }


        public async Task PerformThreadedDistributedTask(
            string CalcName, 
            int BatchSize, 
            Func<int, double> ArgFactory, 
            EventHandler<MathCalcReply> ResponseCallback)
        {
            var state = new DistributedTaskState(CalcName, _channelCollection.Length, BatchSize, ArgFactory, ResponseCallback);

            var t1 = new CancelThread<DistributedTaskState>(Thread1Worker);
            var t2 = new CancelThread<DistributedTaskState>(Thread2Worker);
            t1.Start(state);
            t2.Start(state);

            _ = await state.CompletionSource.Task;
        }


        public void Thread1Worker(DistributedTaskState state, CancellationToken Token)
        {
            int batchID = 0;
            int currentBatchSize = state.BatchSize;
            int currentIdx = 0;

            //continue working until the entire batch is processed
            while (currentIdx < currentBatchSize)
            {
                state.RequestWaitHandle.WaitOne();

                //send batch request to each worker
                //send all batch requests to the workers at once
                //the batches are sent out in waves
                //this allows the response processor to guarantee in-order evaluation

                for (int i = 0; i < _channelCollection.Length && currentIdx < currentBatchSize; ++i)
                {
                    //make a new batch request for this worker
                    var rqstSize = 0;
                    var mathRequest = new MathCalcRequest()
                    {
                        BatchID = batchID++,
                        CalcName = state.CalcName
                    };
                    //Calculate the args for this request
                    //this allows the client to provide an arg factory and lazily evaluate arguments
                    //prevents all arguments from being stored in memory at once
                    for (var j = 0; j < _maxBatchSize && currentIdx < currentBatchSize; ++j, ++currentIdx, ++rqstSize)
                    {
                        var arg = state.ArgFactory(currentIdx);
                        mathRequest.Args.Add(arg);
                    }
                    //dont send request if the batch is empty
                    if (rqstSize == 0)
                    {
                        continue;
                    }
                    //send request to next worker
                    //add awaitable to stack
                    var requestTask = _mathClients[i].DoMathAsync(mathRequest);
                    state.CurrentTaskBatch[i] = requestTask;
                }
                state.ResponseWaitHandle.Set();
            }
        }

        public async void Thread2Worker(DistributedTaskState state, CancellationToken Token)
        {
            //wait for batch responses
            //separates batch dispatching from result callback processing
            //allows one thread to push out new requests while this thread evaluates responses
            //improves performance with non-trivial callbacks

            var cumResponseCount = 0;
            var batchSize = state.BatchSize;

            while (cumResponseCount < batchSize)
            {
                state.ResponseWaitHandle.WaitOne();

                var resultClone = new MathCalcReply[_channelCollection.Length];

                //wait for all worker responses to return before sending new batches
                //process the receipts, then repeat the process
                //cache responses in 2nd buffer
                for (int i = 0; i < _channelCollection.Length; ++i)
                {
                    var task = state.CurrentTaskBatch[i];
                    if (task == null)
                        continue;
                    var response = await task;
                    resultClone[i] = response;
                    state.CurrentTaskBatch[i] = null;
                }
                //let requester thread send new wave of batches
                state.RequestWaitHandle.Set();
                //process response callbacks while new batch wave is being processed by workers
                for (int i = 0; i < _channelCollection.Length; ++i)
                {
                    var response = resultClone[i];
                    if(response != null)
                    {
                        cumResponseCount += response.Responses.Count;
                        state.ResponseCallback?.Invoke(this, response);
                    }
                }
            }
            state.CompletionSource.SetResult(null);
        }



        public async Task PerformDistributedTask(
            string CalcName, 
            int BatchSize, 
            Func<int, double> ArgFactory, 
            EventHandler<MathCalcReply> ResponseCallback)
        {
            var state = new DistributedTaskState(CalcName, _channelCollection.Length, BatchSize, ArgFactory, ResponseCallback);

            int batchID = 0;
            int currentBatchSize = state.BatchSize;
            int currentIdx = 0;

            //continue working until the entire batch is processed
            while (currentIdx < currentBatchSize)
            {
                for (int i = 0; i < _channelCollection.Length; ++i)
                {
                    state.CurrentTaskBatch[i] = null;
                }

                //send batch cluster to each worker
                //send batches to each cluster at once
                for (int i = 0; i < _channelCollection.Length && currentIdx < currentBatchSize; ++i)
                {
                    //make a new batch request for this worker
                    var rqstSize = 0;
                    var mathRequest = new MathCalcRequest()
                    {
                        BatchID = batchID++,
                        CalcName = state.CalcName
                    };
                    //Calculate the args for this request
                    //this allows the client to provide an arg factory and lazily evaluate arguments
                    //prevents all arguments from being stored in memory at once
                    for (var j = 0; j < _maxBatchSize && currentIdx < currentBatchSize; ++j, ++currentIdx, ++rqstSize)
                    {
                        var arg = state.ArgFactory(currentIdx);
                        mathRequest.Args.Add(arg);
                    }
                    //dont send request if the batch is empty
                    if (rqstSize == 0)
                    {
                        continue;
                    }
                    //send request to next worker
                    //add awaitable to stack
                    var requestTask = _mathClients[i].DoMathAsync(mathRequest);
                    state.CurrentTaskBatch[i] = requestTask;
                }

                //wait for all worker responses to return before sending new batches
                //process the receipts, then repeat the process
                for (int i = 0; i < _channelCollection.Length; ++i)
                {
                    var task = state.CurrentTaskBatch[i];
                    if (task == null)
                        continue;
                    var response = await task;
                    state.ResponseCallback?.Invoke(this, response);
                }
            }
        }



        /// <summary>
        /// Shutdown spawned server processes and clients
        /// </summary>
        public void Dispose()
        {
            foreach (var channel in _channelCollection)
            {
                channel.ShutdownAsync().Wait();
            }
            foreach (var proc in _processCollection)
            {
                proc.Dispose();
            }
        }

    }
}

File: Projects/TaskDistributionTestJava/TDT.SpiceEngine/src/main/java/org/Main/WorkerPool.java
package org.Main;

import io.grpc.Channel;
import io.grpc.ManagedChannelBuilder;

import io.grpc.stub.StreamObserver;
import org.Main.Spice.MathCalc.OneArg.MathCalcGrpc.*;
import org.Main.Spice.MathCalc.OneArg.MathCalcGrpc;
import org.Main.Spice.MathCalc.OneArg.MathCalcReply;
import org.Main.Spice.MathCalc.OneArg.MathCalcRequest;
import org.Main.Spice.MathCalc.OneArg.MathCalcRequest.*;
import org.Main.Worker.Thread1Worker;
import org.Main.Worker.Thread2Worker;

import java.io.IOException;
import java.util.function.Function;
import java.util.logging.Logger;


public class WorkerPool  {

    Process[] _processCollection;
    Channel[] _channelCollection;
    MathCalcStub[] _mathClients;
    int _maxBatchSize;
    private static final Logger logger = Logger.getLogger(WorkerPool.class.getName());

    public WorkerPool(String ServerPath, int StartPort, int WorkerCount, int MaxBatchSize) throws IOException {
        int currentPort = StartPort;
        _maxBatchSize = MaxBatchSize;
        _channelCollection = new Channel[WorkerCount];
        _processCollection = new Process[WorkerCount];
        _mathClients = new MathCalcStub[WorkerCount];

        ManagedChannelBuilder<?> channelBuilder;
        for (int i = 0; i < WorkerCount; ++i)
        {
            int port = currentPort++;
            String args = "-p " + port;
            Process proc = Runtime.getRuntime().exec( ServerPath + args); // throws io exception
            _processCollection[i] = proc;

            channelBuilder = getChannelBuilder("127.0.0.1:", port);

            _channelCollection[i] = channelBuilder.build();
            _mathClients[i] = MathCalcGrpc.newStub(_channelCollection[i]);
        }
    }

    private ManagedChannelBuilder<?> getChannelBuilder(String host, int port) {
        return ManagedChannelBuilder.forAddress(host, port).usePlaintext();
    }


    public void PerformThreadedDistributedTask(
            String CalcName,
            int BatchSize,
            Function<Integer, Double> ArgFactory,
            StreamObserver<MathCalcReply> ResponseCallback) // TODO Task replacement
    {
        DistributedTaskState state = new DistributedTaskState(CalcName, _channelCollection.length, BatchSize, ArgFactory, ResponseCallback);

        // TODO: start thread1worker and thread2worker pass in state
        Thread t1 = new Thread(new Thread1Worker(state, _channelCollection, _maxBatchSize, _mathClients));
        Thread t2 = new Thread(new Thread2Worker(state, _channelCollection, _maxBatchSize, _mathClients));
        t1.start();
        t2.start();
    }



    public void PerformDistributedTask(String CalcName, int BatchSize, Function<Integer, Double> ArgFactory, StreamObserver<MathCalcReply> ResponseCallback)
    {
        DistributedTaskState state = new DistributedTaskState(CalcName, _channelCollection.length, BatchSize, ArgFactory,ResponseCallback);

        int batchID = 0;
        int currentBatchSize = state.batchSize;
        int currentIdx = 0;

        //continue working until the entire batch is processed
        while (currentIdx < currentBatchSize)
        {
            for (int i = 0; i < _channelCollection.length; ++i)
            {
                state.currentTaskBatch.set(i, null);
            }

            //send batch cluster to each worker
            //send batches to each cluster at once
            for (int i = 0; i < _channelCollection.length && currentIdx < currentBatchSize; ++i)
            {
                //make a new batch request for this worker
                int rqstSize = 0;
                MathCalcRequest mathRequest = new MathCalcRequest();
                Builder builder = MathCalcRequest.newBuilder().setBatchID(batchID++).setCalcName(state.calcName);
                //Calculate the args for this request
                //this allows the client to provide an arg factory and lazily evaluate arguments
                //prevents all arguments from being stored in memory at once
                for (int j = 0; j < _maxBatchSize && currentIdx < currentBatchSize; ++j, ++currentIdx, ++rqstSize)
                {
                    double arg = state.argFactory.apply(currentIdx);
                    mathRequest = builder.setArgs(currentIdx, arg).build();
                }
                //dont send request if the batch is empty
                if (rqstSize == 0)
                {
                    continue;
                }
                //send request to next worker
                //add awaitable to stack
                StreamObserver<MathCalcReply> requestTask = new StreamObserver<MathCalcReply>() {
                    @Override
                    public void onNext(MathCalcReply value) {

                    }

                    @Override
                    public void onError(Throwable t) {

                    }

                    @Override
                    public void onCompleted() {

                    }
                };
                _mathClients[i].doMath(mathRequest, requestTask);
                state.currentTaskBatch.set(i, requestTask);
            }

            //wait for all worker responses to return before sending new batches
            //process the receipts, then repeat the process
            for (int i = 0; i < _channelCollection.length; ++i)
            {
                StreamObserver<MathCalcReply> task = state.currentTaskBatch.get(i);
                if (task == null)
                    continue;
                StreamObserver<MathCalcReply> response = task; // TODO
                state.responseCallback?.Invoke(this, response); // TODO callback function in replacement
            }
        }
    }

}
